core/base/src/TDirectory.cxx

```diff
@@ -318,11 +318,6 @@ void TDirectory::CleanTargets()
       context->fDirectoryWait = false;
    }
 
-   // Deal with the gROOT case; gROOT does not register the local gDirectories
-   // it is assigned to so the loop over fGDirectories has not done anything.
-   if (this == ROOT::Internal::gROOTLocal && gDirectory == this)
```

Comment:
But aren't you still setting `this` to `ROOT::Internal::gROOTLocal` in some cases? Shouldn't `gDirectory` be pointing to `gROOT` in this case?

Reply:
Yes and it is now handled by the loop on line 288 `      for (auto &ptr : fGDirectories) {` which has also the essential feature of keeping the `shared_ptr` alive until `~TROOT` (indirectly) checks them.